### PR TITLE
WIP: 0.6 Compatibility

### LINKED ISSUE #65: v0.6 precompile error: new UnionAll type breaks things
```
INFO: Precompiling module ColorTypes.
ERROR: LoadError: LoadError: MethodError: no method matching subtypes(::Type{ColorTypes.Color})
Closest candidates are:
  subtypes(::Module, ::DataType) at reflection.jl:316
  subtypes(::DataType) at reflection.jl:333
Stacktrace:
 [1] include_from_node1(::String) at ./loading.jl:532
 [2] include(::String) at ./sysimg.jl:14
 [3] include_from_node1(::String) at ./loading.jl:532
 [4] include(::String) at ./sysimg.jl:14
 [5] anonymous at ./<missing>:2
while loading /home/joe/.julia/v0.6/ColorTypes/src/types.jl, in expression starting on line 417
while loading /home/joe/.julia/v0.6/ColorTypes/src/ColorTypes.jl, in expression starting on line 56
```

can be fixed with:

```
const color3types = filter(x->(!x.abstract && length(fieldnames(x))>1), union(subtypes(Color{T,N}) where T where N, subtypes(AbstractRGB{T}) where T))
```

but then it fails again on 

```
INFO: Precompiling module ColorTypes.
ERROR: LoadError: LoadError: type UnionAll has no field name
Stacktrace:
 [1] macro expansion; at /home/joe/.julia/v0.6/ColorTypes/src/types.jl:555 [inlined]
 [2] anonymous at ./<missing>:?
 [3] include_from_node1(::String) at ./loading.jl:532
 [4] include(::String) at ./sysimg.jl:14
 [5] include_from_node1(::String) at ./loading.jl:532
 [6] include(::String) at ./sysimg.jl:14
 [7] anonymous at ./<missing>:2
while loading /home/joe/.julia/v0.6/ColorTypes/src/types.jl, in expression starting on line 532
while loading /home/joe/.julia/v0.6/ColorTypes/src/ColorTypes.jl, in expression starting on line 56
```

### PR DESCRIPTION
**Pull request by Keno**
Depends on https://github.com/JuliaLang/Compat.jl/pull/304
Fixes #65. Untested on 0.5 so far.

### CODE REVIEW

File: src/ColorTypes.jl [hunk 1]
```diff
@@ -11,6 +11,9 @@ else
     end
 end
 
+import Compat
```
Comment (musm) [inline src/ColorTypes.jl]: do you mean `using`?

**Comment by musm** [comment]
bump

File: src/ColorTypes.jl (hunk 1 above)
Comment (Keno) [inline src/ColorTypes.jl]: no, I'm only using the TypeUtils, no need to create the rest of the bindings

**Comment by tlnagy** [comment]
Triggering build. This should pass now that https://github.com/JuliaLang/METADATA.jl/pull/7828 has gone through.

File: src/types.jl [hunk 2]
```diff
@@ -414,8 +414,16 @@ AGray32(g::AbstractGray, alpha = 1) = AGray32(gray(g), alpha)
 # Note: with the exceptions of `alphacolor` and `coloralpha`, all
 # traits in the rest of this file are intended just for internal use
 
-const color3types = filter(x->(!x.abstract && length(fieldnames(x))>1), union(subtypes(Color), subtypes(AbstractRGB)))
-const parametric3 = filter(x->!isempty(x.parameters), color3types)
+const color3types = map(s->getfield(ColorTypes,s),
+  filter(names(ColorTypes, true)) do s
+    isdefined(ColorTypes, s) || return false
+    t = getfield(ColorTypes, s)
+    isa(t, Type) && t <: Colorant && !isabstract(t) && length(fieldnames(t))>1
```
Comment (timholy) [inline src/types.jl]: julia 0.5:
```julia
julia> using Compat.TypeUtils

julia> using ColorTypes

julia> typealias Color1{T} Color{T,1}
ColorTypes.Color{T,1}

julia> isabstract(Color1)
ERROR: MethodError: no method matching isabstract(::Type{ColorTypes.Color{T,1}})
Closest candidates are:
  isabstract(::DataType) at /home/tim/.julia/v0.5/Compat/src/Compat.jl:1781
```
This causes a precompilation error. Fixed if we add `isabstract(::TypeConstructor) = true` to Compat.

File: src/types.jl [hunk 3]
```diff
@@ -551,7 +559,7 @@ for (C, acol, cola) in [(DIN99d, :ADIN99d, :DIN99dA),
     cfn = Expr(:tuple, colorfields(C)...)
     elty = eltype_default(C)
     ub   = eltype_ub(C)
-    Csym = C.name.name
+    Csym = Base.datatype_name(C)
```
Comment (timholy) [inline src/types.jl]: Missing on julia 0.5.

File: src/traits.jl [hunk 4]
```diff
@@ -121,16 +114,36 @@ For example,
...
+    # dispatches to the first definition below, even when the second is
+    # applicable.
+    _color_type(::Type{TC}) where TC<:(TransparentColor{C, T, N} where T where N) where C = C
+    color_type(::Type{TC}) where TC<:TransparentColor =
+      isa(TC, UnionAll) ? parameter_upper_bound(TC, 1) : _color_type(TC)
+    color_type(::Type{TC}) where TC<:(TransparentColor{C, T, N} where T where N) where C = C
+""")
+else
+    length{T,N}(::Type{Colorant{T,N}}) = N
+    @pure length{C<:Colorant}(::Type{C}) = length(supertype(C))
```
Comment (timholy) [inline src/traits.jl]: ...nassty old typesystem don't like it when hobbitses reorder these methods...

```julia
julia> length(RGB)
ERROR: MethodError: no method matching length(::Type{Any})
Closest candidates are:
  length(::SimpleVector) at essentials.jl:168
  length(::Base.MethodList) at reflection.jl:256
  length(::MethodTable) at reflection.jl:322
  ...
 in length(::Type{ColorTypes.RGB}) at /home/tim/.julia/v0.6/ColorTypes/src/traits.jl:138
```
This package was partially an art in catering to the specific bugs in the older type system.

File: src/traits.jl [hunk 5]
```diff
@@ -151,10 +164,12 @@ base_color_type{C<:Colorant}(::Type{C}) = base_colorant_type(color_type(C))
...
 base_color_type(x::Number)   = Gray
 
-@generated function base_colorant_type{C<:Colorant}(::Type{C})
-    name = C.name.name
-    :($name)
+if isdefined(Core, :UnionAll)
+    basetype(T) = typename(T).wrapper
+else
+    basetype(T) = typename(T).primary
 end
```
Comment (timholy) [inline src/traits.jl]: Breaks inference.

File: src/types.jl [hunk 6]
```diff
@@ -414,8 +414,16 @@ AGray32(g::AbstractGray, alpha = 1) = AGray32(gray(g), alpha)
 # Note: with the exceptions of `alphacolor` and `coloralpha`, all
 # traits in the rest of this file are intended just for internal use
 
-const color3types = filter(x->(!x.abstract && length(fieldnames(x))>1), union(subtypes(Color), subtypes(AbstractRGB)))
-const parametric3 = filter(x->!isempty(x.parameters), color3types)
+const color3types = map(s->getfield(ColorTypes,s),
```
Comment (timholy) [inline src/types.jl]: This block triggers the deprecation warning for `U16` and `U8`. Thoughts?

File: src/types.jl (hunk 2 above)
Comment (timholy) [inline src/types.jl]: https://github.com/JuliaLang/Compat.jl/pull/320

**Comment by timholy** [comment]
See #69.

File: src/traits.jl [hunk 7]
```diff
@@ -237,7 +252,15 @@ ccolor{T,Csrc<:AbstractRGB}(::Type{AbstractRGB{T}}, ::Type{Csrc}) = base_coloran
...
-_ccolor{Cdest,Csrc,T<:Number}(::Type{Cdest}, ::Type{Csrc}, ::Type{T}) = base_colorant_type(Cdest){T}
+if isdefined(Core, :UnionAll)
+include_string("""
+    _ccolor{Cdest,Csrc,T<:Number}(::Type{Cdest}, ::Type{Csrc}, ::Type{T}) = isleaftype(T) ?
+       base_colorant_type(Cdest){T} :
+       base_colorant_type(Cdest){S} where S<:T
+""")
+else
+    _ccolor{Cdest,Csrc,T<:Number}(::Type{Cdest}, ::Type{Csrc}, ::Type{T}) = base_colorant_type(Cdest){T}  
+end
```
Comment (timholy) [inline src/traits.jl]: This is fixed if I `@pure`-ify `basename`. Safe, or not?

Comment (Keno) [inline src/traits.jl]: Should be safe, yes.

File: src/types.jl (hunk 6 above)
Comment (Keno) [inline src/types.jl]: There's a check for something being deprecated, which should be checked first. I thought I had done that. I guess not.

**Pull request closed without merging**